### PR TITLE
Enhanced shard targeting part I 

### PR DESCRIPTION
**Pull request by rafael**
### Description

The following is the first iteration on [enhancing shard targeting](https://github.com/youtube/vitess/issues/3666). Per spec, it adds support to target a range using `keyspace[-20]` range syntax. 

### Implementation notes:

1) `e.resolver.Execute` was having a `keyspace` parameter and it was not being used. I took the opportunity to clean up this.
2) When targeting a single shard, there was no validation that a keyspace was being set. This validation was added.
3) `use` syntax didn't validate that you were providing a valid keyspace. I also added a validation for this.

### TODO for following PR's 
- Add support for this syntax in `FROM` clauses ( i.e ``select a from `keyspace[20-40]`.t``)
- Support for targeting a keyspace_id
- At the moment, targeting a single shard does not benefit from insta commit feature. I maintained this behavior for consistency. However, @demmer  and I were talking that we should do insta commit in these cases. If we decide to do this, I can do this in a different PR.

### CODE REVIEW

**Comment by alainjobart** [comment]
I'm going to sound a bit negative here, but I don't like how the KeyRange is being added to the Target proto message. It is not meant to be used externally over the wire. As far as I can tell, it's only used internally inside vtgate.

Also, the next part of the changes will involve adding keyspace ids, are we also going to add them to the Target structure? I think that would derail this for sure.

The current Target is meant to represent a actual Shard+tablettype: it is reported by vttablet to vtgate as the Shard that tablet belongs to, and its type. I think it will become dangerous and confusing to overload it with anything else. (and for the record, in hindsight, Target is a bad too-generic name).

I think ultimately what we want is a structure that contains the following:
* the keyspace and tablet type
* exactly one of the following:
  * a shard name
  * a keyrange
  * a ksid
  * a list of keyranges
  * a list of ksid

It would be easy to construct it from a Target, but also from parsing you new targetString format. And then we pass it to srvtopo for resolution.

What's funny is @sougou and I were discussing this, and I was going to prototype it soon, because we could use it to communicate between vtgate and srvtopo modules...

Do you mind if I prototype this first (today or tomorrow) and we see if it can be used here?

**Comment by sougou** [comment]
Heh. I didn't expect KsidOrRange to expand into this, but I can see how it can :).

**Comment by rafael** [comment]
@alainjobart agreed. I  thought `Target` was a natural place to extend this, but in light of your comment, I think what you are proposing makes more sense. 

Paraphrasing, just to make sure I'm understanding your idea:
1) We will be able to construct this data structure from `targetString` 
2) We will be able to pass this new data structure to srvtopo and it will resolve the shards.

**Comment by alainjobart** [comment]
Exactly, yes.

And I think eventually, srvtopo will only take that data structure, as it can represent anything we throw at srvtopo right now. So it will actually simplify a lot of things, and remove a lot of code duplication, in vtgate and v3.

**Comment by rafael** [comment]
@alainjobart  - Also, in the short term, so I'm not block, I can remove it from the proto and have an ad-hoc method to parse the range from the targetString. Once we have the new methods, we can refactor just that part to use them.

**Comment by rafael** [comment]
@alainjobart  -  Updated to not have range in `Target` proto.  Let me know your thoughts.

**Comment by alainjobart** [comment]
LGTM

This will work nicely until my change is done.

File: go/vt/vtgate/executor.go [hunk 1]
```diff
@@ -228,18 +236,19 @@ func (e *Executor) handleExec(ctx context.Context, safeSession *SafeSession, sql
...
-
-		execStart := time.Now()
-		logStats.PlanTime = execStart.Sub(logStats.StartTime)
-
+		if keyRange != nil {
+			stmtType := sqlparser.Preview(sql)
+			if stmtType == sqlparser.StmtInsert {
+				return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "range queries not supported for inserts: %s", safeSession.TargetString)
+			}
+			return e.handleRangeExec(ctx, safeSession, sql, bindVars, target, keyRange, logStats)
```
Comment (demmer) [inline go/vt/vtgate/executor.go]: This early return doesn't update the logStats properly which it does in the shard-targeted branch.

File: go/vt/vtgate/executor.go [hunk 2]
```diff
@@ -283,25 +292,50 @@ func (e *Executor) handleExec(ctx context.Context, safeSession *SafeSession, sql
 	return qr, err
 }
 
+func (e *Executor) handleRangeExec(ctx context.Context, safeSession *SafeSession, sql string, bindVars map[string]*querypb.BindVariable, target querypb.Target, keyRange *topodatapb.KeyRange, logStats *LogStats) (*sqltypes.Result, error) {
```
Comment (demmer) [inline go/vt/vtgate/executor.go]: For consistency with `shardExec` I feel like this should be called `shardRangeExec`

File: go/vt/vtgate/executor.go (hunk 1 above)
Comment (rafael) [reply go/vt/vtgate/executor.go]: Good catch. Updating.

Comment (rafael) [reply go/vt/vtgate/executor.go]: Actually, I think it is. It's done inside the method. Am I missing one?

Comment (demmer) [reply go/vt/vtgate/executor.go]: Oh right. Well given the asymmetry with the shardExec branch this feels like something is missing.

I'd say we should either move the logStats updates into `shardExec` or pull them out of `shardRangeExec` but doing both seems confusing.

Comment (rafael) [reply go/vt/vtgate/executor.go]: Done. While doing this, I realized there was a place missing in the validation return. Fixed.

File: go/vt/vtgate/executor.go (hunk 2 above)
Comment (rafael) [reply go/vt/vtgate/executor.go]: Done.

**Comment by sougou** [comment]
Now that I'm looking at this code, I see what the problem is. This approach, and the upcoming design change that @alainjobart is working on, are at conflict.

But before I come up with a good suggestion, I have one question for @alainjobart: how will retries be handled in V3? Let's say we get an "unavailable" error, which could be caused by a resharding event, or by local vttablets going unavailable. Then the v3 code must re-resolve, which will result in a different set of targets.

Basically, I think that's the level at which TargetString should be resolved.

**Comment by alainjobart** [comment]
I think everything will be pushed down to the resolver layer. We'll pass in the TargetString and let the Resolver convert into []*resolvedShards, and it should just work. I'll have a PR coming later today with the data structures I was mentioning...

**Comment by sougou** [comment]
The problem I see, I don't think I've explained it very well. I'll gather my thoughts and phrase them better by today evening.

**Comment by sougou** [comment]
Also, this PR can still be submitted. The concerns I have are likely things that have to be fixed forward.

**Comment by alainjobart** [comment]
I meant to say the call to the resolver / TargetString parser will be done inside the 'f func' that is passed down to the executor side, and will be called again to re-resolve, if needed. So in case of retry, we'd re-resolve from scratch.

**Comment by alainjobart** [comment]
Here is my PR that introduces key.Destination.
https://github.com/youtube/vitess/pull/3699
But as we said, somewhat orthogonal at this point, we can merge this one and fix forward.

File: go/vt/vtgate/executor.go [hunk 3]
```diff
@@ -284,24 +300,38 @@ func (e *Executor) handleExec(ctx context.Context, safeSession *SafeSession, sql
...
 	if target.Keyspace == "" {
 		return nil, errNoKeyspace
 	}
+	keyRange, err := parseRange(safeSession.TargetString)
+	if err != nil {
+		return nil, errNoKeyspace
+	}
 
 	f := func() ([]*srvtopo.ResolvedShard, error) {
 		var destination key.Destination
```
Comment (alainjobart) [inline go/vt/vtgate/executor.go]: These six lines can be taken out of the closure, and right after the previous block:
keyRange, err := ...
...
var destination key.Destination
if keyRange != nil {
...

f := func() (....)

File: go/vt/vtgate/executor.go [hunk 4]
```diff
@@ -284,24 +300,38 @@ func (e *Executor) handleExec(ctx context.Context, safeSession *SafeSession, sql
 	return qr, err
 }
 
-func (e *Executor) shardExec(ctx context.Context, safeSession *SafeSession, sql string, bindVars map[string]*querypb.BindVariable, target querypb.Target, logStats *LogStats) (*sqltypes.Result, error) {
+func (e *Executor) destinationExec(ctx context.Context, safeSession *SafeSession, sql string, bindVars map[string]*querypb.BindVariable, target querypb.Target, destination key.Destination, logStats *LogStats) (*sqltypes.Result, error) {
 	f := func() ([]*srvtopo.ResolvedShard, error) {
-		return e.resolver.resolver.ResolveDestination(ctx, target.Keyspace, target.TabletType, key.DestinationShard(target.Shard))
+		var result []*srvtopo.ResolvedShard
```
Comment (alainjobart) [inline go/vt/vtgate/executor.go]: Please name this variable rss, for consistency.

Also, can just be:
rss, err := ...

File: go/vt/vtgate/executor.go [hunk 5]
```diff
@@ -229,18 +244,19 @@ func (e *Executor) handleExec(ctx context.Context, safeSession *SafeSession, sql
...
 			sql = normalized + comments
 		}
-
+		logStats.PlanTime = execStart.Sub(logStats.StartTime)
 		logStats.SQL = sql
 		logStats.BindVariables = bindVars
 
-		execStart := time.Now()
-		logStats.PlanTime = execStart.Sub(logStats.StartTime)
+		var destination key.Destination
```
Comment (alainjobart) [inline go/vt/vtgate/executor.go]: These few lines can be merged with the 'if keyRange != nil' statement on line 228, so it's all done in one place.

Comment (rafael) [reply go/vt/vtgate/executor.go]: Done. Funny, I did it like that first, but then changed it. Thought it was better to separate validation from this logic.

File: go/vt/vtgate/executor.go (hunk 3 above)
Comment (rafael) [reply go/vt/vtgate/executor.go]: Done. When I did this, I also notice that then this can also re-use `e.destinationExec`

File: go/vt/vtgate/executor.go (hunk 4 above)
Comment (rafael) [reply go/vt/vtgate/executor.go]: Done.

**Comment by rafael** [comment]
Just squashed everything. I'd say, once travis finished, good to merge from my perspective.
